packages/fairmq/package.py

```diff
@@ -9,74 +12,77 @@ class Fairmq(CMakePackage):
-    depends_on('msgpack-c@3.1.1', when='@1.4.3')
-    depends_on('msgpack-c@3.1.1', when='@dev')
-
-    depends_on('dds@2.1-1-g181b66a', when='@1.2.3')
-    depends_on('dds@2.2', when='@1.4.3')
-    depends_on('dds@master', when='@dev')
-
-    depends_on('nanomsg@1.0.0', when='@1.2.3')
-    depends_on('nanomsg@1.1.5', when='@1.4.3')
-    depends_on('nanomsg@1.1.5', when='@dev')
-
-    depends_on('flatbuffers', when='@dev')
+    maintainers = ['dennisklein', 'ChristianTackeGSI']
+    generator = 'Ninja'
+
+    version('develop', branch='dev', submodules=True, get_full_repo=True)
+    version('1.4.12', tag='v1.4.12', commit='9f8a3553ba7e6c3ce5c9e378d77d3e16d7ce6daf', submodules=True, no_cache=True)
+    version('1.4.11', tag='v1.4.11', commit='e6dede492e02d6966cfa68e6820f90f679b74592', submodules=True, no_cache=True)
+    version('1.4.10', tag='v1.4.10', commit='f31be6d7a10ab0b7e4183576438268c8122117a5', submodules=True, no_cache=True)
+    version('1.4.9', tag='v1.4.9', commit='f6e3183f453d3daf97e529d3b834da715b43c47b', submodules=True, no_cache=True)
+    version('1.4.8', tag='v1.4.8', commit='3f5374820a56c472a79e81b977861eddf4bff343', no_cache=True)
+    version('1.4.7', tag='v1.4.7', commit='7cacf471b957aa1a89616b84339ce0b25699c7a0', no_cache=True)
+    version('1.4.6', tag='v1.4.6', commit='93eb599df7d9ee4c869e2fe70c03b7a845739683', no_cache=True)
+    version('1.4.5', tag='v1.4.5', commit='b1c82641237c1da7ae4d7af1fcb1cf69f596a54c', no_cache=True)
+    version('1.4.4', tag='v1.4.4', commit='99c8d33191a8f69bcd6b51cfa8c11a1d736734b0', no_cache=True)
+    version('1.4.3', tag='v1.4.3', commit='3582091b1c76cbe689b8e66813ea71dd5ff55bc5', no_cache=True)
+    version('1.4.2', tag='v1.4.2', commit='2457094b6c4004a75620a45329574d983cdc263e', no_cache=True)
+    version('1.4.1', tag='v1.4.1', commit='95ec56dcf0d95917a0c9cdecd1004d4a1c4aa479', no_cache=True)
+    version('1.4.0', tag='v1.4.0', commit='4c2785dfc1d98576917bc3cfb2a01eab675c5f6e', no_cache=True)
+    version('1.3.9', tag='v1.3.9', commit='d91a7d23616d2781938423b2362651185e2c0607', no_cache=True)
+    # TODO Once https://github.com/spack/spack/issues/14344 is resolved, enable
+    #      source caching again (by removing the `no_cache` argument).
+
+    variant('build_type', default='RelWithDebInfo',
+            values=('Debug', 'Release', 'RelWithDebInfo'),
+            multi=False,
+            description='CMake build type')
+    variant('cxxstd', default='default',
+            values=('11', '14', '17'),
+            multi=False,
+            description='Force the specified C++ standard when building.')
+    conflicts('cxxstd=11', when='@1.4.11:')
+
+    patch('fix_find_dds.patch', when='@1.4.0:1.4.4')
+
+    depends_on('googletest@1.7:', when='@:1.4.8')
+    depends_on('boost@1.64: +container+program_options+thread+system+filesystem+regex+date_time', when='@1.3')
+    depends_on('boost@1.64: +container+program_options+filesystem+date_time+regex', when='@1.4')
```

Comment:
Where is the boost dependency for `@develop`?
I start to wonder, why the unit test works for `@develop`.
Looking at the last cdash output, boost is missing from the `fairmq@develop` concretization… strange.

Fix is easy: turn `1.4` into `1.4:`. But I'd like to understand, why the unit test works.

Reply:
> But I'd like to understand, why the unit test works.

By default, the CMake command `find_package` is trying to be very smart about finding a suitable dependency, see https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure. My suspicion is that even though spack does not provide an explicit hint, CMake is still able to find a suitable Boost. I will investigate and add the necessary flags to switch off this behaviour.

Reply:
Turns out the explanation is even simpler: Even though Boost is not concretized as a direct dependency of FairMQ, spack still puts it into the `$CMAKE_INSTALL_PREFIX`. My guess is that the `setup_dependent_build_environment` of transitive dependencies is not filtered out (because the spack dependency model has no knowledge about which of them are public dependencies - they have to leave them in).